src/mctp-bench.c

```diff
@@ -0,0 +1,413 @@
+        if (recv_ctx->stats.curr_packet_len > MAX_LEN) {
+                warn("recv: expected max len:%zd bytes, got:%zd bytes",
+                     MAX_LEN, recv_ctx->stats.curr_packet_len);
+                return -1;
+        }
+
+        hdr = (struct msg_header *)recv_ctx->buf;
+        if (hdr->magic != MAGIC_VAL) {
+                warnx("recv: expected magic:\"%x\", got:\"%x\"\n", MAGIC_VAL,
+                      hdr->magic);
+                return -1;
+        }
+
+        recv_ctx->stats.total_received_len += recv_ctx->stats.curr_packet_len;
+        recv_ctx->stats.msg_count++;
+
+        if (!valid_payload(recv_ctx->buf, recv_ctx->stats.curr_packet_len))
+                recv_ctx->stats.invalid_payloads++;
+
+        if (!recv_ctx->started_recv_flag) {
+                printf("recv: first msg received\n");
+                recv_ctx->started_recv_flag = true;
+                clock_gettime(CLOCK_MONOTONIC, &recv_ctx->start_time);
+                recv_ctx->stats.prev_seq_no = hdr->seq_no;
+                recv_ctx->stats.msgs_dropped +=
+                    get_packets_dropped(hdr->seq_no, SEQ_START);
+                return -1;
+        }
+
+        recv_ctx->stats.msgs_dropped +=
+            get_packets_dropped(hdr->seq_no, recv_ctx->stats.prev_seq_no);
+
+        recv_ctx->stats.prev_seq_no = hdr->seq_no;
+        return 0;
+}
+
+static int mctp_bench_recv()
+{
+        struct recv_ctx recv_ctx = {0};
+        struct sockaddr_mctp addr = {0};
+        int rc;
+
+        recv_ctx.sd = socket(AF_MCTP, SOCK_DGRAM, 0);
+        if (recv_ctx.sd < 0)
+                err(EXIT_FAILURE, "recv: socket");
+
+        addr.smctp_family = AF_MCTP;
+        addr.smctp_network = MCTP_NET_ANY;
+        addr.smctp_addr.s_addr = MCTP_ADDR_ANY;
+        addr.smctp_type = 1;
+        addr.smctp_tag = MCTP_TAG_OWNER;
+
+        recv_ctx.buf = malloc(MAX_LEN);
+        if (!recv_ctx.buf) {
+                err(EXIT_FAILURE, "recv: malloc failed");
+        }
+
+        rc = bind(recv_ctx.sd, (struct sockaddr *)&addr, sizeof(addr));
+        if (rc) {
+                close(recv_ctx.sd);
+                err(EXIT_FAILURE, "recv: bind failed");
+        }
+
+        recv_ctx.started_recv_flag = false;
+
+        printf("recv: waiting for first msg\n");
+        while (1) {
+                int timeout;
+                struct pollfd pollfd;
+                pollfd.fd = recv_ctx.sd;
+                pollfd.events = POLLIN;
+
+                if (recv_ctx.started_recv_flag) {
+                        clock_gettime(CLOCK_MONOTONIC, &recv_ctx.current_time);
+                        timeout = get_timeout(&recv_ctx);
+                } else
+                        timeout = -1;
+
+                rc = poll(&pollfd, 1, timeout);
+                if (rc < 0) {
+                        warn("recv: poll failed");
+                        break;
+                }
+
+                if (pollfd.revents & POLLIN) {
```

Comment:
I think it should probably test `rc == 1` from `poll`? (Not certain what happens if `poll()` times out, whether it clears `revents` or not)

Reply:
Applied suggested change, now it won't attempt to receive and handle the message if rc is not 1.